File: src/main/Controllers/TextController.py
import requests
import json



class TextController:
    def __init__(self):
        # Base URL for Flask microservice
        # self.base_url = "http://localhost:5001/api/extract"
        self.base_url = "http://209.38.252.155:5001/api/extract"
    
    def convert_to_text(self, pdf_data):
        # Prepare the file data as part of multipart form-data request
        files = {
            'file': ('uploaded.pdf', pdf_data, 'application/pdf')
        }

        try:
            # Send POST request to Flask microservice with the PDF data
            response = requests.post(self.base_url, files=files)
            
            # Check if the response was successful
            if response.status_code == 200:
                # Parse JSON response
                response_data = response.json()
                # Return the extracted text from the response
                return response_data.get('text', 'No text extracted')
            else:
                # Handle unsuccessful response
                return f"Error: {response.status_code} - {response.text}"
        except Exception as ex:
            # Handle exception (e.g., connection errors)
            return f"Exception: {str(ex)}"

# Example usage:
if __name__ == '__main__':
    # Initialize TextController
    controller = TextController()
    
    # Read the PDF file to be sent
    with open("/Users/taha/Desktop/001_Taha/002_Studium/002_Uni/001_Fank_UAS/Uni/5.Semester/NachSchreib/Dist_Sys/Lectures/07_blockchains.pdf", "rb") as pdf_file:
        pdf_data = pdf_file.read()
    
    # Get extracted text
    extracted_text = controller.convert_to_text(pdf_data)
    
    # Print the extracted text
    print(extracted_text)

File: LernAssistent/Controllers/TextController.cs
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading.Tasks;
using System.IO;
using LernAssistent.Interfaces;
using Newtonsoft.Json;

namespace LernAssistent.Controllers
{
    public class TextController
    {
        private static readonly HttpClient _httpClient = new HttpClient();

        // Constructor
        public TextController()
        {
            _httpClient.BaseAddress = new Uri("http://localhost:5001"); // Flask microservice URL
        }

        // Method to send PDF data to the Flask microservice and get the extracted text
        public async Task<string> ConvertToText(byte[] pdfData)
        {
            using (var content = new MultipartFormDataContent())
            {
                // Create the file content and add it to the request
                var fileContent = new ByteArrayContent(pdfData);
                fileContent.Headers.ContentType = MediaTypeHeaderValue.Parse("application/pdf");
                content.Add(fileContent, "file", "uploaded.pdf");

                try
                {
                    // Send the request to the Flask microservice
                    var response = await _httpClient.PostAsync("/api/extract", content);

                    // Check if the response is successful
                    if (response.IsSuccessStatusCode)
                    {
                        // Read the response content
                        var responseContent = await response.Content.ReadAsStringAsync();

                        // Deserialize JSON response to extract text
                        var result = JsonConvert.DeserializeObject<Response>(responseContent);
                        return result?.Text;
                    }
                    else
                    {
                        // Handle unsuccessful response
                        var errorContent = await response.Content.ReadAsStringAsync();
                        return $"Error: {errorContent}";
                    }
                }
                catch (Exception ex)
                {
                    return $"Exception: {ex.Message}";
                }
            }
        }

        // Response class to deserialize JSON response
        public class Response
        {
            public string Text { get; set; }
        }
    }
}
